bundle: be more resilient against timeouts

philikon opened the pull request:
Also don't attempt to fetch the bundle until the RN packager server is actually up. There's little point in trying sooner.

philikon committed:
bundle: wait up to 30s for the RN packager server to come up

Also don't attempt to fetch the bundle until the RN packager server is actually up. There's little point in trying sooner.

jhabdas commented on lib/Server.js:
```diff
@@ -130,12 +130,13 @@ class Server {
           resolve();
         });
       });
+      return Promise.all([listenPromise, readyPromise]);
```
`Server.start  is called by `program` (`commander`) in `/bin/rnws.js`. Based on https://github.com/tj/commander.js/issues/352 returning a `Promise` here isn't doing anything anyway, so I don't see a lot of value in this change.

philikon replied:
I'm not quite sure I understand the second sentence of your comment.

philikon replied:
The reason for having `Server.start()` not resolve its promise until all servers are up and running is so that the HTTP request made by `createBundle()` doesn't time out while the RN packager and webpack are still traversing the file system for their dependencies. On slow machines this can take quite a while. With this change, the HTTP request doesn't get made until both packagers are ready to receive requests.

jhabdas replied:
My understanding is the Promise will return immediately. And unless the
async/await pattern, or generator functions are used, the net result will
be the same WRT timing of the two servers starting. I could be wrong, but
it seems the only functional change is the timeout, which I'd argue is the
responsibility of the CI system.
On Fri, Nov 27, 2015 at 9:01 PM Philipp von Weitershausen <
notifications@github.com> wrote:

> In lib/Server.js
> <https://github.com/mjohnston/react-native-webpack-server/pull/144#discussion_r46075231>
> :
>
> >          this.httpServer = this.server.listen(this.port, () => {
> >            console.log(`Server listening at http://${this.hostname}:${this.port}`);
> >            resolve();
> >          });
> >        });
> > +      return Promise.all([listenPromise, readyPromise]);
>
> The reason for having Server.start() not resolve its promise until all
> servers are up and running is so that the HTTP request made by
> createBundle() doesn't time out while the RN packager and webpack are
> still traversing the file system for their dependencies. On slow machines
> this can take quite a while. With this change, the HTTP request doesn't get
> made until both packagers are ready to receive requests.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mjohnston/react-native-webpack-server/pull/144/files#r46075231>
> .
>

philikon replied:
> My understanding is the Promise will return immediately. And unless the async/await pattern, or generator functions are used, the net result will be the same WRT timing of the two servers starting. 

You seem to be misunderstanding how promises work. There's a difference between _returning_ a promise (which usually happens immediately) and _resolving_ the promise (which usually happens whenever you'd invoke a callback indicating success of an asynchronous action).

When constrained to ES5 idioms, this is how you'd use a promise:

``` js
somethingThatReturnsAPromise().then(function (result) {
  doSomethingWith(result);
});
```

An equivalent, more modern usage of promises would be this:

``` js
let result = await somethingThatReturnsAPromise();
doSomethingWith(result);
```

Functionally, both ways are 100% equivalent.

You can see for yourself that the implementation for `rnws bundle` uses the first variation: https://github.com/mjohnston/react-native-webpack-server/blob/168e0809cf14ade8aaf756bb95ec484fa386c2f8/bin/rnws.js#L146-L147. As you can see, this makes the HTTP request as soon as the `server.start()` promise resolves. My change is to make the `server.start()` promise resolve when both the RN packager _and_ the webpack server are ready to accept requests. Because if we open the request sooner and the RN packager takes a really long time to start up, we might risk timing out the incoming request because we can't respond to it soon enough.

> but it seems the only functional change is the timeout, which I'd argue is the responsibility of the CI system.

How do you propose to do that? In what way can the CI system take responsibility here?

jhabdas replied:
Through config. But I bow out here. Good luck.
On Fri, Nov 27, 2015 at 11:47 PM Philipp von Weitershausen <
notifications@github.com> wrote:

> In lib/Server.js
> <https://github.com/mjohnston/react-native-webpack-server/pull/144#discussion_r46076347>
> :
>
> >          this.httpServer = this.server.listen(this.port, () => {
> >            console.log(`Server listening at http://${this.hostname}:${this.port}`);
> >            resolve();
> >          });
> >        });
> > +      return Promise.all([listenPromise, readyPromise]);
>
> My understanding is the Promise will return immediately. And unless the
> async/await pattern, or generator functions are used, the net result will
> be the same WRT timing of the two servers starting.
>
> You seem to be misunderstanding how promises work. There's a difference
> between *returning* a promise (which usually happens immediately) and
> *resolving* the promise (which usually happens whenever you'd invoke a
> callback indicating success of an asynchronous action).
>
> When constrained to ES5 idioms, this is how you'd use a promise:
>
> somethingThatReturnsAPromise().then(function (result) {
>   doSomethingWith(result);
> });
>
> An equivalent, more modern usage of promises would be this:
>
> let result = await somethingThatReturnsAPromise();doSomethingWith(result);
>
> Functionally, both ways are 100% equivalent.
>
> You can see for yourself that the implementation for rnws bundle uses the
> first variation:
> https://github.com/mjohnston/react-native-webpack-server/blob/168e0809cf14ade8aaf756bb95ec484fa386c2f8/bin/rnws.js#L146-L147.
> As you can see, this makes the HTTP request as soon as the server.start()
> promise resolves. My change is to make the server.start() promise resolve
> when both the RN packager *and* the webpack server are ready to accept
> requests. Because if we open the request sooner and the RN packager takes a
> really long time to start up, we might risk timing out the incoming request
> because we can't respond to it soon enough.
>
> but it seems the only functional change is the timeout, which I'd argue is
> the responsibility of the CI system.
>
> How do you propose to do that? In what way can the CI system take
> responsibility here?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mjohnston/react-native-webpack-server/pull/144/files#r46076347>
> .
>

jhabdas commented on lib/Server.js:
```diff
@@ -283,7 +284,10 @@ class Server {
+      // waitForSocket retries the port every 250ms. Let's give the
+      // React Native server up to 30 seconds to come online. watchman
+      // can be slow to ramp up, especially on CI machines.
+      waitForSocket({ port: this.packagerPort, tries: 120 }, err => {
```
Which CI machines?

philikon replied:
Travis.

ghost commented:
CI build failed for this change failed. Ironically due to a timeout.

philikon commented:
> Ironically due to a timeout.

Yeah, that _is_ pretty funny. I wonder if the `Promise.all()` change is responsible for that somehow...

philikon committed:
Explicitly terminate the node process after the bundle has been written and the server has been shut down

This is to avoid having the node process potentially wait forever to be killed, which is more likely to happen now that we explicitly wait for all servers to be up before making the HTTP request(s) for the bundle(s).

philikon committed:
Disable timeouts for incoming and outgoing HTTP requests

philikon commented:
Being cautiously optimistic here, but these changes seem to have fixed a slew of long-standing issues with the super slow Travis CI machines. It'd be great to get them merged and released.

@elliottsj @mjohnston, I know you guys aren't using this actively, but would either of you be willing to review this code here and spin a release? Alternatively, I'll be happy to take care of it too (minus the review part of course) if you guys want to grant me push/release privileges.

mjohnston commented:
Change looks good to me. I'll get you set up with push/release access.

philikon commented:
Thanks @mjohnston!

mjohnston commented:
No problem - you should also be able to publish the npm package now.

philikon commented:
Sweet! I'll release 0.8.2 soon.
